docs/source/cpp/memory.rst

```diff
@@ -201,3 +201,261 @@ simply do::
    std::shared_ptr<arrow::Buffer> arbitrary_buffer = ... ;
    std::shared_ptr<arrow::Buffer> cpu_buffer = arrow::Buffer::ViewOrCopy(
       arbitrary_buffer, arrow::default_cpu_memory_manager());
+
+
+Memory Profiling
+================
+
+On Linux, detailed profiles of memory allocations can be generated using perf record,
+without any need to modify the binaries. These profiles can show the
+traceback in addition to allocation parameters (like size).
+
+.. TODO: This requires a debug build, right? Or maybe not if symbols in header file
+
+
+.. note::
+   If you profiling Arrow's tests on another platform, you can run the following docker container
+   using archery:::
+
+      archery docker run ubuntu-cpp bash
+      /arrow/ci/scripts/cpp_build.sh /arrow /build
+      cd build/cpp/debug
+      ./arrow-array-test # Run a test
+      apt-get update
+      apt-get install -y linux-tools-generic
+      alias perf=/usr/lib/linux-tools/<something>/perf
+
+
+To track allocations, create probe points on each of the jemalloc methods used.
+Collecting ``'$params'`` allows us to record the size of the allocations requested,
+while collecting ``$retval`` allows us to record the address of recorded allocations,
+so we can correlate them with the call to free/dealloc.
+
+:: 
+
+   perf probe -x libarrow.so je_arrow_mallocx '$params' 
+   perf probe -x libarrow.so je_arrow_mallocx%return '$retval' 
+   perf probe -x libarrow.so je_arrow_rallocx '$params' 
+   perf probe -x libarrow.so je_arrow_rallocx%return '$retval' 
+   perf probe -x libarrow.so je_arrow_dallocx '$params' 
+
+Then you can record calls with associated tracebacks using ``perf record``. In this 
+example, we are running the StructArray unit tests in Arrow::
+   
+   perf record -g --call-graph dwarf \
+    -e probe_libarrow:je_arrow_mallocx \
+    -e probe_libarrow:je_arrow_mallocx__return \
+    -e probe_libarrow:je_arrow_rallocx \
+    -e probe_libarrow:je_arrow_rallocx__return \
+    -e probe_libarrow:je_arrow_dallocx \
+    ./arrow-array-test --gtest_filter=StructArray*
+
+.. TODO: What are the equivalent probe calls for mimalloc and system allocator?
```

Comment:
They should all be in https://github.com/apache/arrow/blob/master/cpp/src/arrow/memory_pool.cc